### PR TITLE
bump PyTorch version to 1.11

### PR DESCRIPTION
**Pull request by t-vi**
This bumps PyTorch to 1.11 and fixes 3 test failures. The bump is required to enable the libtorch_ops fallback due to DLPack version incompatibilities.

QAT training has its own `fuse_modules` version (`fuse_modules_qat`) in PyTorch, so I changed the test.

Two amendments to the front end:
- `searchsorted` gets more (optional) parameters to its signature,
- There is a `sub` variant with `alpha` (`a - alpha * b`). PyTorch rewrites `rsub` with `alpha` to this, but we ignored it. Now we handle `sub` with alpha.

Thank you,  @masahi for getting me started with the bump and pointing out the test failures. Any errors are my own.

### CODE REVIEW

**Comment by t-vi** [comment]
Caffe2 has been dropped from PyTorch, which is what we are getting here: https://github.com/pytorch/pytorch/issues/67151

**Comment by t-vi** [comment]
I think I could need a hint how to proceed, given that this is likely a major issue for all caffe2 use in TVM.

**Comment by masahi** [comment]
Unless there is a standalone way to install caffe2, then I think removing caffe2 support entirely is the only way forward. Our caffe2 frontend hasn't been updated for years, so I don't think people would object...

**Comment by masahi** [comment]
We can propose dropping caffe2 support to the community next week. In the meantime, we can remove caffe2 bits from CI:
* Remove https://github.com/apache/tvm/blob/main/docker/Dockerfile.ci_gpu#L80-L81
* Remove https://github.com/apache/tvm/blob/92a80e93446331780cfa81452b164862845e7c65/tests/scripts/task_python_frontend.sh#L60
* Probably need to remove https://github.com/apache/tvm/blob/44549e623433dd10d9e97e442ef529fb44c46c14/gallery/how_to/compile_models/from_caffe2.py

**Comment by masahi** [comment]
Also `ci-qemu` has failed to install PT 1.11 for some version conflict reason https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/detail/PR-10794/1/pipeline/67. But I don't know why it needs to install pytorch... probably it only needs onnx. So I suggest decoupling pytorch install from `ubuntu_install_onnx.sh` and create `ubuntu_install_pytorch.sh`.

File: python/tvm/relay/frontend/pytorch.py
Comment (masahi) [inline python/tvm/relay/frontend/pytorch.py:2951]: It seems this change breaks `test_lstm.py`:
```
FAILED test_lstm.py::test_custom_lstm - AttributeError: 'function' object has no attribute 'dtype'
```

Comment (t-vi) [reply python/tvm/relay/frontend/pytorch.py:2951]: I'll see to fixing it. Thank you.

**Comment by t-vi** [comment]
> Also ci-qemu has failed to install PT 1.11 for some version conflict reason

Ohoh. I think this is because Python 3.6 is EOL upstream and PyTorch doesn't support it anymore...
This is a larger can of worms than I had hoped for. :slightly_smiling_face:

**Comment by masahi** [comment]
I thought we are now running Python 3.7, see
* https://github.com/apache/tvm/pull/10247
* https://github.com/apache/tvm/pull/10654

but maybe the qemu image hasn't been updated yet.

**Comment by leandron** [comment]
> I thought we are now running Python 3.7, see
> 
> * [[docker] Update CI to Python 3.7 and Ubuntu 18 #10247](https://github.com/apache/tvm/pull/10247)
> * [[CI] Update docker images to bring Python 3.7 and TensorFlow 2.6 #10654](https://github.com/apache/tvm/pull/10654)
> 
> but maybe the qemu image hasn't been updated yet.

On this, it looks like ci_qemu (https://github.com/apache/tvm/blob/main/docker/Dockerfile.ci_qemu) is not installing Python using the commons script: https://github.com/apache/tvm/blob/main/docker/install/ubuntu1804_install_python.sh.

This probably needs to be fixed in a separate PR. Do you want to send that fix? (Asking just because I won’t be able to take this for the next ~two weeks).

**Comment by masahi** [comment]
Note that probably we need to update the ci image version in https://github.com/apache/tvm/blob/main/Jenkinsfile#L54 as well. For that, we need to wait until a new nightly image containing the update from your previous PR is pushed to https://hub.docker.com/r/tlcpackstaging/ci_qemu/tags. 

So please wait another day before resuming PT 1.11 update work, or remove pytorch install from `ubuntu_install_onnx.sh`.

**Comment by t-vi** [comment]
@masahi Thank you for merging the qemu Python bump and the advice. I think waiting for it to show up might be the cleanest option, especially given that I still need to learn so much about how the CI works. :)

**Comment by t-vi** [comment]
@masahi OK, so now we're part of the nightly with the qemu update. How would I get a version tag that is useful for bumping the version in the Jenkinsfile?

**Comment by masahi** [comment]
Yeah, so the way it works is as follows

* We update the `ci-docker-staging` branch https://github.com/apache/tvm/tree/ci-docker-staging to point `Jenkinsfile` to the new image in `tlcpackstaging`. 
* After the CI passes, we retag the nightly image as `v0.83` etc and push the image to `tlcpack` dockerhub. 
* Update `Jenkinsfile` in `main` to point to the new image on `tlcpack`, and send a PR to merge this change.

Note that the above process needs to happen for every CI image update. Right now we are in the middle of `ci-qemu` update, but after we merge this PR to update PT, we need to go through the same exercises to update `ci-gpu`.

I've been trying to run a CI job on `ci-docker-staging`, I keep getting various errors but hopefully I can pass it in a few hours.

**Comment by t-vi** [comment]
Thank you @masahi . So if I understand this right, the next step is something you need to do? I'd appreciate a shout if I can proceed here or help other bits along.

**Comment by masahi** [comment]
Yes, pushing changes to `ci-docker-staging`, or pushing a new image to `tlcpack` dockerhub need to be done by a committer. No worry, I've done this many times. But since we need to wait for at least two CI runs (one for `ci-docker-staging` to test the new image, another for `main` to actually update the image), things won't be done by today.

**Comment by leandron** [comment]
> Yes, pushing changes to `ci-docker-staging`, or pushing a new image to `tlcpack` dockerhub need to be done by a committer. No worry, I've done this many times. But since we need to wait for at least two CI runs (one for `ci-docker-staging` to test the new image, another for `main` to actually update the image), things won't be done by today.

Just a heads up that you're likely to see the issue with updated containers - https://github.com/apache/tvm/issues/10696.

**Comment by masahi** [comment]
Yeah I hit that error once a couple of hours before, fortunately the ongoing run https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/detail/ci-docker-staging/238/pipeline/ didn't hit it.

**Comment by masahi** [comment]
@t-vi the PR https://github.com/apache/tvm/pull/10814 should unblock this work.

**Comment by masahi** [comment]
https://github.com/apache/tvm/pull/10815

**Comment by masahi** [comment]
https://github.com/apache/tvm/pull/10815 was merged so we can resume the work on this PR. I restarted a CI job by closing / reopening (there should be a better way than this...)

**Comment by t-vi** [comment]
`[2022-03-29T23:00:39.045Z] unknown parent image ID sha256:f75815a47f249990da41ca0e349ede16f8710bd2d573de74a9afbd1a9b528055`

in the docker build. I would not even know what to look at here... :confused:

**Comment by masahi** [comment]
Hopefully it is just a flaky issue, since the error came from the unrelated image `ci-hexagon`

**Comment by t-vi** [comment]
I'm not sure whether it is flakiness or merging main, but it seems to be past that bit now. Hopefully, if there are more failures, it'll be something I can look into to fix. Thank you for all your help @masahi !

**Comment by masahi** [comment]
you got an error in one of tests because the CI environment is still using PT 1.10. We get to use 1.11 at the last step of https://github.com/apache/tvm/pull/10794#issuecomment-1081579377

So can you revert that change or use different code paths depending on versions?

**Comment by t-vi** [comment]
I think I have the error from a premature 1.12 compat change, I'm fixing this right now but want to test locally.

**Comment by masahi** [comment]
Another `unknown parent image ID` error https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/detail/PR-10794/6/pipeline
cc @driazati (the other one was https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/detail/PR-10794/4/pipeline/)

**Comment by t-vi** [comment]
All green. :slightly_smiling_face:

**Review by masahi — APPROVED** [review]

**Comment by t-vi** [comment]
Thank you, @masahi for merging and helping me. So next we would need to update the docker gpu image used before I can return to enabling the test that needs PyTorch 1.11?

**Comment by masahi** [comment]
Yes, the next step is to wait until the nightly image appears in https://hub.docker.com/r/tlcpackstaging/ci_gpu/tags. It will happen about 12h later.

**Comment by driazati** [comment]
I started a manual build so we don't have to wait: https://ci.tlcpack.ai/job/docker-images-ci/job/daily-docker-image-rebuild/272/

Usually takes about an hour

**Comment by leandron** [comment]
To confirm here, as I did the update this time... the latest version of the images now contains PyTorch 1.11.

```
$ docker run -it --rm tlcpack/ci-gpu:v0.84 bash
root@6ba934f13b82:/# python3
Python 3.7.5 (default, Dec  9 2021, 17:04:37) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.__version__
'1.11.0+cu102'
>>> 
```

**Comment by masahi** [comment]
The update to PT 1.11 was already done in https://github.com/apache/tvm/pull/10849

**Pull request merged**
